File: src/sautice/invoice/roster.py
"""Workspace roster: customers and a product catalogue, with name resolution.

Resolution never "snaps" to the nearest name. A single clear match resolves; a
tie inside a confusable cluster (Adebayo Stores vs Adebayo Ventures) stays
ambiguous and becomes a question; nothing recognisable becomes a candidate new
entity. The roster is demonstration data — real workspaces are loaded the same way.
"""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

# repo_root/benchmarks/data/sautibench/roster.json ; this file is src/sautice/invoice/roster.py
DEFAULT_ROSTER = Path(__file__).resolve().parents[3] / "benchmarks" / "data" / "sautibench" / "roster.json"


def load_roster(path: str | Path | None = None) -> dict[str, Any]:
    p = Path(path) if path else DEFAULT_ROSTER
    return json.loads(p.read_text(encoding="utf-8"))


def _norm(s: str) -> str:
    return re.sub(r"\s+", " ", re.sub(r"[^a-z0-9 ]", " ", s.lower())).strip()


def _tokens(s: str) -> set[str]:
    # Drop tokens that carry no identity, so "Musa" matches "Musa Hardware".
    stop = {"stores", "store", "ltd", "limited", "nig", "nigeria", "co", "company",
            "enterprises", "ventures", "trading", "global", "concept", "services",
            "and", "sons", "brothers", "&"}
    return {t for t in _norm(s).split() if t and t not in stop}


def resolve_customer(query: str, customers: list[dict]) -> dict[str, Any]:
    q = _norm(query)
    if not q:
        return {"status": "missing", "query": query, "candidates": []}

    exact = [c for c in customers if _norm(c["name"]) == q]
    if len(exact) == 1:
        return {"status": "resolved", "query": query,
                "resolved": _cust(exact[0]), "candidates": [_cust(exact[0])]}

    qt = _tokens(query)
    hits = []
    for c in customers:
        name_n = _norm(c["name"])
        overlap = len(qt & _tokens(c["name"]))
        if q and q in name_n:
            hits.append((overlap + 5, c))
        elif overlap:
            hits.append((overlap, c))
    hits.sort(key=lambda x: -x[0])

    if not hits:
        return {"status": "unknown", "query": query, "candidates": []}
    top = hits[0][0]
    tied = [c for s, c in hits if s == top]
    if len(tied) == 1:
        return {"status": "resolved", "query": query,
                "resolved": _cust(tied[0]), "candidates": [_cust(c) for _, c in hits[:5]]}
    return {"status": "ambiguous", "query": query, "candidates": [_cust(c) for _, c in hits[:5]]}


def resolve_product(query: str, products: list[dict]) -> dict[str, Any]:
    q = _norm(query)
    if not q:
        return {"status": "missing", "query": query, "candidates": []}

    exact = [p for p in products if _norm(p["name"]) == q]
    if len(exact) == 1:
        return {"status": "resolved", "query": query,
                "resolved": _prod(exact[0]), "candidates": [_prod(exact[0])]}

    qt = _tokens_prod(query)
    hits = []
    for p in products:
        names = [p["name"], *p.get("aliases", [])]
        alias_hit = any(_norm(a) in q or q in _norm(a) for a in p.get("aliases", []))
        overlap = max((len(qt & _tokens_prod(n)) for n in names), default=0)
        if alias_hit:
            hits.append((overlap + 5, p))
        elif overlap:
            hits.append((overlap, p))
    hits.sort(key=lambda x: -x[0])

    if not hits:
        return {"status": "unknown", "query": query, "candidates": []}
    top = hits[0][0]
    tied = [p for s, p in hits if s == top]
    if len(tied) == 1:
        return {"status": "resolved", "query": query,
                "resolved": _prod(tied[0]), "candidates": [_prod(p) for _, p in hits[:5]]}
    return {"status": "ambiguous", "query": query, "candidates": [_prod(p) for _, p in hits[:5]]}


def _tokens_prod(s: str) -> set[str]:
    stop = {"the", "a", "of", "and"}
    return {t for t in _norm(s).split() if t and t not in stop}


def _cust(c: dict) -> dict:
    return {"id": c["id"], "name": c["name"]}


def _prod(p: dict) -> dict:
    return {"id": p["id"], "name": p["name"], "unit": p.get("unit"),
            "unit_price_kobo": p["unit_price_kobo"]}


def find_customer(cid: str, customers: list[dict]) -> dict | None:
    for c in customers:
        if c["id"] == cid:
            return _cust(c)
    return None


def find_product(pid: str, products: list[dict]) -> dict | None:
    for p in products:
        if p["id"] == pid:
            return _prod(p)
    return None

File: web/src/lib/invoice/core/roster.ts
/**
 * Workspace roster + name resolution — ported from src/sautice/invoice/roster.py.
 * A single clear match resolves; a confusable-cluster tie stays ambiguous (a
 * question); an unrecognised name becomes a candidate new entity — never snapped.
 *
 * roster.json here is a synced copy of benchmarks/data/sautibench/roster.json
 * (bundled so the serverless function needs no filesystem read).
 */
import rosterData from "../data/roster.json";

export interface Customer { id: string; name: string; phone?: string; cluster?: string }
export interface Product { id: string; name: string; unit?: string; unit_price_kobo: number; aliases?: string[]; cluster?: string }
export interface Roster { business?: Record<string, unknown>; customers: Customer[]; products: Product[] }

export const ROSTER = rosterData as unknown as Roster;

export interface Resolution<T> { status: "resolved" | "ambiguous" | "unknown" | "missing"; query: string; resolved?: T; candidates: T[] }

const norm = (s: string) => s.toLowerCase().replace(/[^a-z0-9 ]/g, " ").replace(/\s+/g, " ").trim();

const CUST_STOP = new Set(["stores", "store", "ltd", "limited", "nig", "nigeria", "co", "company",
  "enterprises", "ventures", "trading", "global", "concept", "services", "and", "sons", "brothers", "&"]);
const PROD_STOP = new Set(["the", "a", "of", "and"]);

const toks = (s: string, stop: Set<string>) => new Set(norm(s).split(" ").filter((t) => t && !stop.has(t)));
const overlap = (a: Set<string>, b: Set<string>) => [...a].filter((x) => b.has(x)).length;

const custView = (c: Customer): Customer => ({ id: c.id, name: c.name });
const prodView = (p: Product): Product => ({ id: p.id, name: p.name, unit: p.unit, unit_price_kobo: p.unit_price_kobo });

export function resolveCustomer(query: string, customers: Customer[]): Resolution<Customer> {
  const q = norm(query);
  if (!q) return { status: "missing", query, candidates: [] };

  const exact = customers.filter((c) => norm(c.name) === q);
  if (exact.length === 1) return { status: "resolved", query, resolved: custView(exact[0]), candidates: [custView(exact[0])] };

  const qt = toks(query, CUST_STOP);
  const hits: [number, Customer][] = [];
  for (const c of customers) {
    const ov = overlap(qt, toks(c.name, CUST_STOP));
    if (q && norm(c.name).includes(q)) hits.push([ov + 5, c]);
    else if (ov) hits.push([ov, c]);
  }
  hits.sort((a, b) => b[0] - a[0]);
  if (hits.length === 0) return { status: "unknown", query, candidates: [] };
  const top = hits[0][0];
  const tied = hits.filter(([s]) => s === top);
  if (tied.length === 1) return { status: "resolved", query, resolved: custView(tied[0][1]), candidates: hits.slice(0, 5).map(([, c]) => custView(c)) };
  return { status: "ambiguous", query, candidates: hits.slice(0, 5).map(([, c]) => custView(c)) };
}

export function resolveProduct(query: string, products: Product[]): Resolution<Product> {
  const q = norm(query);
  if (!q) return { status: "missing", query, candidates: [] };

  const exact = products.filter((p) => norm(p.name) === q);
  if (exact.length === 1) return { status: "resolved", query, resolved: prodView(exact[0]), candidates: [prodView(exact[0])] };

  const qt = toks(query, PROD_STOP);
  const hits: [number, Product][] = [];
  for (const p of products) {
    const names = [p.name, ...(p.aliases ?? [])];
    const aliasHit = (p.aliases ?? []).some((a) => norm(a).includes(q) || q.includes(norm(a)));
    const ov = Math.max(0, ...names.map((n) => overlap(qt, toks(n, PROD_STOP))));
    if (aliasHit) hits.push([ov + 5, p]);
    else if (ov) hits.push([ov, p]);
  }
  hits.sort((a, b) => b[0] - a[0]);
  if (hits.length === 0) return { status: "unknown", query, candidates: [] };
  const top = hits[0][0];
  const tied = hits.filter(([s]) => s === top);
  if (tied.length === 1) return { status: "resolved", query, resolved: prodView(tied[0][1]), candidates: hits.slice(0, 5).map(([, p]) => prodView(p)) };
  return { status: "ambiguous", query, candidates: hits.slice(0, 5).map(([, p]) => prodView(p)) };
}

export const findCustomer = (id: string, customers: Customer[]): Customer | null => {
  const c = customers.find((x) => x.id === id);
  return c ? custView(c) : null;
};
export const findProduct = (id: string, products: Product[]): Product | null => {
  const p = products.find((x) => x.id === id);
  return p ? prodView(p) : null;
};
